subprojects/logging/src/main/java/org/gradle/internal/logging/sink/GroupingProgressLogEventGenerator.java

```diff
@@ -142,46 +174,49 @@ private Object getOperationId(@Nullable final Object buildOpId) {
         private final String category;
         private final String loggingHeader;
         private final long startTime;
+        private long lastUpdateTime;
 
         private List<RenderableOutputEvent> bufferedLogs = new ArrayList<RenderableOutputEvent>();
 
         private OperationGroup(String category, @Nullable String loggingHeader, long startTime, Object buildOpIdentifier) {
             this.category = category;
             this.loggingHeader = loggingHeader;
             this.startTime = startTime;
+            this.lastUpdateTime = startTime;
             this.buildOpIdentifier = buildOpIdentifier;
         }
 
         private LogEvent spacerLine() {
-            return new LogEvent(startTime, category, null, "", null);
+            return new LogEvent(lastUpdateTime, category, null, "", null);
         }
 
         StyledTextOutputEvent header(final String message) {
             List<StyledTextOutputEvent.Span> spans = Lists.newArrayList(new StyledTextOutputEvent.Span(StyledTextOutput.Style.Header, "> " + message), new StyledTextOutputEvent.Span(EOL));
-            return new StyledTextOutputEvent(startTime, category, null, buildOpIdentifier, spans);
+            return new StyledTextOutputEvent(lastUpdateTime, category, null, buildOpIdentifier, spans);
         }
 
         void bufferOutput(RenderableOutputEvent output) {
-            bufferedLogs.add(output);
+            // Forward output immediately when the focus is on this operation group
+            if (Objects.equal(buildOpIdentifier, lastRenderedBuildOpId)) {
+                listener.onOutput(output);
+                lastUpdateTime = timeProvider.getCurrentTime();
+            } else {
+                bufferedLogs.add(output);
+            }
         }
 
         void flushOutput() {
```

Comment:
Something seems a little fishy here.  Since `flushOutput` happens on a separate thread, I don't think there are any guarantees about how `bufferedLogs`, `lastRenderedBuildOpId` and `lastUpdateTime` are synchronized.   So we could potentially render all buffered logs (in the "flush" thread), add more (in a separate thread) and then clear it (in the flush thread).  Which would lose output.